src/Core/Picking.js

```diff
@@ -0,0 +1,159 @@
+
+const fakeScene = new THREE.Scene();
+
+function hideEverythingElse(view, object, threejsLayer = 0) {
+    // We want to render only 'object' and its hierarchy.
+    // So we:
+    //    - place it in a fake scene
+    //    - if it uses threejsLayer defined -> force it on the camera
+    //      (or use the default one: 0)
+    const oldParent = object.parent;
+    const oldScene = view.scene;
+    const prev = view.camera.camera3D.layers.mask;
+
+    fakeScene.add(object);
+
+    view.scene = fakeScene;
+
+    view.camera.camera3D.layers.mask = 1 << threejsLayer;
+
+    return () => {
+        view.scene = oldScene;
+        oldParent.add(object);
+        view.camera.camera3D.layers.mask = prev;
+    };
+}
+
+// TileMesh picking support function
+function screenCoordsToNodeId(view, tileLayer, mouse) {
+    const dim = view.mainLoop.gfxEngine.getWindowSize();
+
+    mouse = mouse || new THREE.Vector2(Math.floor(dim.x / 2), Math.floor(dim.y / 2));
+
+    const restore = tileLayer.level0Nodes.map(n => n.pushRenderState(RendererConstant.ID));
+
+    const undoHide = hideEverythingElse(view, tileLayer.object3d, tileLayer.threejsLayer);
+
+    var buffer = view.mainLoop.gfxEngine.renderViewTobuffer(
+        view,
+        view.mainLoop.gfxEngine.fullSizeRenderTarget,
+        mouse.x, dim.y - mouse.y,
+        1, 1);
+
+    undoHide();
+
+    restore.forEach(r => r());
+
+    var depthRGBA = new THREE.Vector4().fromArray(buffer).divideScalar(255.0);
+
+    // unpack RGBA to float
+    var unpack = unpack1K(depthRGBA, Math.pow(256, 3));
+
+    return Math.round(unpack);
+}
+
+function findLayerIdInParent(obj) {
+    if (obj.layer) {
+        return obj.layer;
+    }
+    if (obj.parent) {
+        return findLayerIdInParent(obj.parent);
+    }
+}
+
+const raycaster = new THREE.Raycaster();
+
+export default {
+    pickTilesAt: (_view, mouse, layer) => {
+        const results = [];
+        const _id = screenCoordsToNodeId(_view, layer, mouse);
+
+        const extractResult = (node) => {
+            if (node.id === _id && node instanceof TileMesh) {
+                results.push({
+                    object: node,
+                    layer: layer.id,
+                });
+            }
+        };
+        for (const n of layer.level0Nodes) {
+            n.traverse(extractResult);
+        }
+        return results;
+    },
+
+    pickPointsAt: (view, mouse, layer) => {
+        if (!layer.root) {
+            return;
+        }
+
+        const dim = view.mainLoop.gfxEngine.getWindowSize();
+
+        // enable picking mode for points material
+        layer.object3d.traverse((o) => {
+            if (o.isPoints && o.baseId) {
+                o.material.enablePicking(true);
+            }
+        });
+
+        const undoHide = hideEverythingElse(view, layer.object3d, layer.threejsLayer);
+
+        // render 1 pixel
+        // TODO: support more than 1 pixel selection
+        const buffer = view.mainLoop.gfxEngine.renderViewTobuffer(
```

Comment:
may be change function parameters to `renderViewTobuffer(scene, camera, x, y, width, height)` to clarify the scene switch in the viewer (and avoid `fakeScene`)

Reply:
Let's discuss the change to this function in the dedicated PR: https://github.com/iTowns/itowns/pull/672

In the meantime this PR can use the fakeScene trick because it's harmless.

Reply:
I simplified this code and removed the need for the `fakeScene`